fix(compiler): work around TypeScript bug when narrowing switch statements

crisbeto opened the pull request:
We type check `@switch` blocks by generating identical TS `switch` statements in the TCB, however TS currently has a bug where parenthesized `switch` block expressions don't narrow their types. Since we use parenthesized expressions to wrap AST nodes for diagnostics, this will bug will affect all Angular-generated `switch` statements.

These changes work around the issue by generating `if`/`else if`/`else` statements that represent the `switch`.

Some alternatives that were considered:
1. Moving the `switch` expression to a constant - this is fairly simple to implement, but it won't fully resolve the narrowing issue since the same constant will have to be used in expressions inside the different cases.
2. Removing the outer-most parenthesis from the switch expression - this works and allows us to continue using switch statements, but because we use parenthesized expressions to map diagnostics to their template locations, I wasn't sure if it won't lead to worse template dignostics.

Fixes #52077.

cexbrayat commented:
@crisbeto I wanted to bring to your attention https://github.com/angular/angular/issues/52107 where I was asking about the possibility of having exhaustive switches in templates. If the TCB generates `if` statements, will that be possible to achieve?

JoostK commented:
> @crisbeto I wanted to bring to your attention #52107 where I was asking about the possibility of having exhaustive switches in templates. If the TCB generates `if` statements, will that be possible to achieve?

Yep, no problem:

```ts
if (x.type === 'foo') {}
else if (x.type === 'bar') {}
else {
  const _: never = x.type;
}
```

will report an error if `x.type` has not been narrowed to the `never` type. I'm not sure whether the reported error message would be clear enough though.

cexbrayat commented:
@JoostK Ha yes, the `never` trick 👍 I hope this lands one way or another, I think it would be a great feature. Thanks for the answer!

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -1273,27 +1273,47 @@ class TcbSwitchOp extends TcbOp {
-            tcbExpression(current.expression, this.tcb, clauseScope),
-            [...clauseScope.render(), breakStatement]));
+  private generateCase(
+      index: number, switchValue: ts.Expression,
```
Since the `switchValue` expression is now emitted multiple times, we'd need to use `markIgnoreDiagnostics` for all "secondary" usages to only report errors in the switch value once.

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -1273,27 +1273,47 @@ class TcbSwitchOp extends TcbOp {
-      const clauseScope = Scope.forNodes(this.tcb, this.scope, null, current.children, null);
+    const expression = tcbExpression(this.block.expression, this.tcb, this.scope);
+    const root = this.generateCase(0, expression, null);
+    root && this.scope.addStatement(root);
```
Compiler uses strict-null equality
```suggestion
    if (root !== null) {
      this.scope.addStatement(root);
    }
```

crisbeto committed:
fix(compiler): work around TypeScript bug when narrowing switch statements

We type check `@switch` blocks by generating identical TS `switch` statements in the TCB, however TS currently has a bug where parenthesized `switch` block expressions don't narrow their types. Since we use parenthesized expressions to wrap AST nodes for diagnostics, this will bug will affect all Angular-generated `switch` statements.

These changes work around the issue by generating `if`/`else if`/`else` statements that represent the `switch`.

Some alternatives that were considered:
1. Moving the `switch` expression to a constant - this is fairly simple to implement, but it won't fully resolve the narrowing issue since the same constant will have to be used in expressions inside the different cases.
2. Removing the outer-most parenthesis from the switch expression - this works and allows us to continue using switch statements, but because we use parenthesized expressions to map diagnostics to their template locations, I wasn't sure if it won't lead to worse template dignostics.

Fixes #52077.

crisbeto commented:
Reworked based on Joost's feedback.

atscott commented:
This PR was merged into the repository by commit 386e1e950033ad98661e5077a4f119df0e7b3008.
